src/Open3D/Visualization/Visualizer/GuiVisualizer.cpp

```diff
@@ -1764,68 +1765,107 @@ bool GuiVisualizer::SetIBL(const char *path) {
     return result;
 }
 
-bool GuiVisualizer::LoadGeometry(const std::string &path) {
-    auto geometry = std::shared_ptr<geometry::Geometry3D>();
+void GuiVisualizer::LoadGeometry(const std::string &path) {
+    auto progressbar = std::make_shared<gui::ProgressBar>();
+    gui::Application::GetInstance().PostToMainThread(this, [this, path,
+                                                            progressbar]() {
+        auto &theme = GetTheme();
+        auto loading_dlg = std::make_shared<gui::Dialog>("Loading");
+        auto vert =
+                std::make_shared<gui::Vert>(0, gui::Margins(theme.font_size));
+        auto loading_text = std::string("Loading ") + path;
+        vert->AddChild(std::make_shared<gui::Label>(loading_text.c_str()));
+        vert->AddFixed(theme.font_size);
+        vert->AddChild(progressbar);
+        loading_dlg->AddChild(vert);
+        ShowDialog(loading_dlg);
+    });
 
-    auto geometry_type = io::ReadFileGeometryType(path);
+    gui::Application::GetInstance().RunInThread([this, path, progressbar]() {
+        auto UpdateProgress = [this, progressbar](float value) {
+            gui::Application::GetInstance().PostToMainThread(
+                    this,
+                    [progressbar, value]() { progressbar->SetValue(value); });
```

Comment:
This will work, but some loaders will call update_progress on every point they load, potentially millions of times before GUI gets to process these events (and while GUI processes those millions of events, it will hold posted_lock_ which will block UpdateProgress which will block the loader, making it wait for GUI to finish before it can continue loading).  Ideally ProgressBar would only post only 1 PostToMainThread call only to flag PostRedraw on the window, and until it gets redrawn it would only update value (atomically).